### PR TITLE
 [ownership] Initialize owner configuration

### PR DESCRIPTION
**Pull request by cfrantz**
Depending on the `ownership_state`, initialize the owner configuration
appropriately.
- Locked states initialize the chip with a single configuration and make
  sure the two owner pages are identical.
- Unlocked states initialize the chip with an A/B configuration based on
  the contents of each of the owner pages.
- The No-Owner state normally performs no initialization.  However, a
  weak function is introduced to allow FPGA builds to link a `testonly`
  module that supplies the default testing owner configuration.

### CODE REVIEW

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 1]
```diff
@@ -243,9 +250,16 @@ OT_WARN_UNUSED_RESULT
 static rom_error_t rom_ext_verify(const manifest_t *manifest,
                                   const boot_data_t *boot_data) {
   RETURN_IF_ERROR(rom_ext_boot_policy_manifest_check(manifest, boot_data));
-  const sigverify_rsa_key_t *key;
-  RETURN_IF_ERROR(sigverify_rsa_key_get(
-      sigverify_rsa_key_id_get(&manifest->rsa_modulus), &key));
+  // const sigverify_rsa_key_t *key;
+  // RETURN_IF_ERROR(sigverify_rsa_key_get(
+  //     sigverify_rsa_key_id_get(&manifest->rsa_modulus), &key));
```
Comment (timothytrippel) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:255]: can this just be deleted?

**Review by timothytrippel — APPROVED** [review]

File: sw/device/silicon_creator/rom_ext/rom_ext.c (hunk 1 above)
Comment (cfrantz) [reply sw/device/silicon_creator/rom_ext/rom_ext.c:255]: Done.

File: sw/device/silicon_creator/lib/ownership/ownership.c [hunk 2]
```diff
@@ -1,21 +1,177 @@
...
+        sizeof(owner_page[1]) / sizeof(uint32_t), &owner_page[1]));
+    owner_page_valid[1] = owner_page_valid[0];
+  } else if (owner_page_valid[0] != kOwnerPageStatusSealed &&
+             owner_page_valid[1] != kOwnerPageStatusSealed) {
+    // Neither page is valid; go to the LockedNone state.
+    dbg_printf("error: both owner pages invalid.\r\n");
+    bootdata->ownership_state = kOwnershipStateLockedNone;
+    nonce_new(&bootdata->nonce);
+    HARDENED_RETURN_IF_ERROR(boot_data_write(bootdata));
+    return kErrorOwnershipBadInfoPage;
```
Comment (moidx) [inline sw/device/silicon_creator/lib/ownership/ownership.c:87]: What do you think about `s/kOwnershipStateLockedNone/kOwnershipStateRecovery/`.

However, is there a way to recover here? It feels that the only mechanism available would be to bootstrap an image to initialize the chip. We could also include a recovery key in the ROM_EXT to support owner recovery if bootstrap is not available...

**Review by moidx — APPROVED** [review]

File: sw/device/silicon_creator/lib/ownership/owner_block.h [hunk 3]
```diff
@@ -38,6 +38,16 @@ typedef struct owner_application_keyring {
   const owner_application_key_t *key[16];
 } owner_application_keyring_t;
 
+/**
+ * Initialize the owner config with default values.
+ *
+ * The sram_exec mdoe is set to DisabledLocked and the three configuration
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/owner_block.h:44]: Fixed.

File: sw/device/silicon_creator/lib/ownership/ownership.c [hunk 4]
```diff
@@ -1,21 +1,177 @@
...
 #include "sw/device/silicon_creator/lib/ownership/ecdsa.h"
+#include "sw/device/silicon_creator/lib/ownership/owner_block.h"
+#include "sw/device/silicon_creator/lib/ownership/ownership.h"
 #include "sw/device/silicon_creator/lib/ownership/ownership_key.h"
 
 // RAM copy of the owner INFO pages from flash.
 owner_block_t owner_page[2];
+owner_page_status_t owner_page_valid[2];
+
+static owner_page_status_t check_owner_page_validity(size_t page) {
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership.c:24]: Fixed (I think it's better as `owner_page_validity_check`, but I think `static` functions don't have to follow the naming convention of externally visible symbols).

File: sw/device/silicon_creator/lib/ownership/ownership.c [hunk 5]
```diff
@@ -1,21 +1,177 @@
...
+  // This really needs to be the KMAC over the page with a keymgr-created
+  // sealing secret.
+  hmac_digest_t digest;
+  hmac_sha256(&owner_page[page], seal_len, &digest);
+  if (hardened_memeq(owner_page[page].seal, digest.digest,
+                     ARRAYSIZE(digest.digest)) == kHardenedBoolTrue) {
+    return kOwnerPageStatusSealed;
+  }
+  hardened_bool_t result = ownership_key_validate(
+      /*page=*/page, kOwnershipKeyOwner, &owner_page[page].signature,
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership.c:40]: Fixed.

File: sw/device/silicon_creator/lib/ownership/ownership.c [hunk 6]
```diff
@@ -1,21 +1,177 @@
...
+      &owner_page[page], sig_len);
+  if (result == kHardenedBoolFalse) {
+    // If the page is bad, destroy the RAM copy.
+    memset(&owner_page[page], 0x5a, sizeof(owner_page[0]));
+  }
+  return result == kHardenedBoolTrue ? kOwnerPageStatusSigned
+                                     : kOwnerPageStatusInvalid;
+}
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership.c:48]: Fixed.

File: sw/device/silicon_creator/lib/ownership/ownership_activate.c [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/ownership/ecdsa.h"
+#include "sw/device/silicon_creator/lib/ownership/ownership.h"
+#include "sw/device/silicon_creator/lib/ownership/ownership_key.h"
+
+static rom_error_t activate(boot_svc_msg_t *msg, boot_data_t *bootdata) {
+  size_t len = (uintptr_t)&msg->ownership_activate_req.signature -
+               (uintptr_t)&msg->ownership_activate_req.primary_bl0_slot;
+  // First check if page1 is even in a valid state for a transfor.
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership_activate.c:20]: Fixed.

File: sw/device/silicon_creator/lib/ownership/ownership_activate.c [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    return kErrorOwnershipInvalidNonce;
+  }
+
+  // Seal page one to this chip.
+  ownership_page_seal(/*page=*/1);
+
+  // Program the sealed page into slot 1.
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(&kFlashCtrlInfoPageOwnerSlot1,
+                                                 kFlashCtrlEraseTypePage));
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership_activate.c:42]: I added a TODO for that.

File: sw/device/silicon_creator/lib/ownership/ownership.c (hunk 2 above)
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/ownership.c]: Renamed `LockedNone` to `Recovery`.

The `silicon_creator` has the option to compile into the ROM_EXT a `kNoOwnerRecoveryKey`.  If that key is present in the ROM_EXT, then that key can be used with the ownership unlock command to transition the chip from `OwnershipStateLockedNone` to one of the unlocked states (`Unlocked{Any, Endorsed}`) and then re-apply an ownership configuration.

If a chip finds its way to the `LockedNone` state, the options then are:
1. Use the No-owner recovery key and the rescue protocol to recover the chip.
2. SPI-bootstrap a special ROM_EXT to forcibly recover the chip.

Both operations involve `silicon_creator` unless `silicon_creator` delegates use of the No-owner recovery key to another entity.

**Pull request merged**